File: back/src/domain/models/captcha.rs
pub struct Captcha {
    raw: String,
}

impl Captcha {
    #[cfg(test)]
    pub fn test() -> Self {
        Self {
            raw: String::from("123456"),
        }
    }
}

impl From<Captcha> for String {
    fn from(captcha: Captcha) -> Self {
        captcha.raw
    }
}

impl TryFrom<String> for Captcha {
    type Error = ();

    fn try_from(raw: String) -> Result<Self, Self::Error> {
        if raw.is_empty() {
            Err(())
        } else {
            Ok(Self { raw })
        }
    }
}

File: src/domain/models/captcha.ts
import { Result } from "@utils";

export type Captcha = {
    _type: "captcha";
    raw: string;
};

export type CaptchaError = "empty";

export const Captcha = {
    generate(): Captcha {
        return {
            _type: "captcha",
            raw: "123456"
        };
    },
    toString(captcha: Captcha): string {
        return captcha.raw;
    },
    tryFromString(captcha: string): Result<Captcha, CaptchaError> {
        const raw = captcha.trim();
        if (raw.length === 0) { return Result.err("empty"); }
        return Result.ok({ _type: "captcha", raw });
    }
};
